[Merge-Queue] Extract bug_id when updating pull-request

JonWBedard opened the pull request:
#### 2207fa9fc0025529b5c8821513f9e541f671e931
<pre>
[Merge-Queue] Extract bug_id when updating pull-request
<a href="https://bugs.webkit.org/show_bug.cgi?id=238772">https://bugs.webkit.org/show_bug.cgi?id=238772</a>
&lt;rdar://problem/91263398 &gt;

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/steps.py:
(LeaveComment.start): Leave comment on both bugzilla and pull request.
(LeaveComment.getResultSummary): Ditto.
(UpdatePullRequest):
(UpdatePullRequest.bug_id_from_log): Extract bug_id from commit message.
(UpdatePullRequest.evaluateCommand): Set bug_id property
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/249268@main">https://commits.webkit.org/249268@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@292405">https://svn.webkit.org/repository/webkit/trunk@292405</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1764,11 +1764,12 @@ def start(self):
-        elif self.bug_id:
-            rc = self.comment_on_bug(self.bug_id, self.comment_text)
-        else:
+            rc += SUCCESS if self.comment_on_pr(self.pr_number, self.comment_text) else FAILURE
```
let's not do arithmetric operations on rc (it can be very hard for others to follow and need knowledge of what's the underlying values of SUCCESS, FAILURE etc). 
Let's just directly assign the value it should be.

JonWBedard replied:
While that's fine for the first one, it's not possible for the second. We want this step to both comment on an associated PR and the associated bugzilla, if both properties are available. This isn't something we want to do in general, but is something we want to do when landing (which is the motivation of this change)

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1778,19 +1779,25 @@ def start(self):
-            return buildstep.BuildStep.getResultSummary(self)
+                summary += f'Added comment on PR {self.pr_number}'
+            if self.bug_id:
+                summary += f'{" and a" if summary else "A"}dded comment on bug {self.bug_id}'
```
this seems convoluted to read, can we have explicit if else

JonWBedard replied:
We could, but we will duplicate both the PR and bug comment summaries

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1778,19 +1779,25 @@ def start(self):
+            if self.pr_number:
+                summary += f'Failed to add comment on PR {self.pr_number}'
+            if self.bug_id:
+                summary += f'{" and f" if summary else "F"}ailed to add comment on bug {self.bug_id}'
```
Ditto

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -4944,13 +4958,27 @@ def evaluateCommand(self, cmd):
         title = loglines[4][4:].rstrip()
         description = f'#### {loglines[0].split()[1]}\n<pre>\n'
         for line in loglines[4:]:
+            if not bug_id:
```
this should be a separate method

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -4944,13 +4958,27 @@ def evaluateCommand(self, cmd):
         description += '</pre>\n'
 
+        if bug_id:
+            self.setProperty('bug_id', bug_id, 'commit')
```
Can have more descriptive name than 'commit', maybe the step name, or can just leave it to default ( i guess default is the step name).

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1764,11 +1764,12 @@ def start(self):
-            rc = self.comment_on_bug(self.bug_id, self.comment_text)
-        else:
+        if self.bug_id:
+            rc = rc if self.comment_on_bug(self.bug_id, self.comment_text) == SUCCESS else FAILURE
```
Maybe can simplify it something like:
```
if self.bug_id and self.comment_on_bug(self.bug_id, self.comment_text) != SUCCESS:
    rc= FAILURE
```

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -4940,6 +4953,17 @@ def getResultSummary(self):
+                    match = candidate.match(word)
+                    if match:
+                        return match.group('id')
+
```
Nit: can remove this empty line

JonWBedard committed:
[Merge-Queue] Extract bug_id when updating pull-request
https://bugs.webkit.org/show_bug.cgi?id=238772
<rdar://problem/91263398>

Patch by Jonathan Bedard <jbedard@apple.com> on 2022-04-05
Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/steps.py:
(LeaveComment.start): Leave comment on both bugzilla and pull request.
(LeaveComment.getResultSummary): Ditto.
(UpdatePullRequest):
(UpdatePullRequest.bug_id_from_log): Extract bug_id from commit message.
(UpdatePullRequest.evaluateCommand): Set bug_id property
* Tools/CISupport/ews-build/steps_unittest.py:
